### PR TITLE
Add `@hasOneThrough` directive

### PR DESCRIPTION
**Pull request by saeed-rostami**
https://github.com/nuwave/lighthouse/issues/332
 
- [x] Added or updated tests
- [x] Documented user facing changes

**Changes**

Hello, since I noticed that the Has one through relationship is not in the document, I decided to add it.
I tried to write the example in the doc similar to the Laravel doc.
tests are added and passed successfully on my local machine.
thanks

**Breaking changes**

### CODE REVIEW

File: src/Schema/Directives/HasOneThroughDirective.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+"""
+Corresponds to [the Eloquent relationship HasOneThrough](https://laravel.com/docs/eloquent-relationships#has-one-through).
+"""
+directive @hasOneThrough(
+  """
+  Specify the relationship method name in the model class,
+  if it is named different from the field in the schema.
+  """
+  relation: String
+  """
```
Comment (spawnia) [inline src/Schema/Directives/HasOneThroughDirective.php:21]:
```suggestion

  """
```

File: tests/Integration/Schema/Directives/HasOneThroughDirectiveTest.php
Comment (spawnia) [inline tests/Integration/Schema/Directives/HasOneThroughDirectiveTest.php:1]: Please clean up the whitespace in this file (no double blank lines, indentation, etc.)

**Review by spawnia — CHANGES_REQUESTED** [review]
Please try to use the existing test models instead of adding new ones. The addition of 3 new test models, factories and migrations adds quite a bit of maintenance burden.

File: tests/Utils/Models/Post.php [hunk 2]
```diff
@@ -102,4 +103,11 @@ public function roles(): HasMany
     {
         return $this->hasMany(RoleUserPivot::class, 'user_id', 'user_id');
     }
+
+
```
Comment (spawnia) [inline tests/Utils/Models/Post.php:107]: no double newlines

File: tests/Utils/Models/Post.php [hunk 3]
```diff
@@ -102,4 +103,11 @@ public function roles(): HasMany
...
+    /** @return \Illuminate\Database\Eloquent\Relations\HasOne<PostStatus> */
```
Comment (spawnia) [inline tests/Utils/Models/Post.php:108]: always use fqcn in docblocks

File: tests/Utils/Models/PostStatus.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ *   Foreign keys
+ * @property int|null $post_id
+ *
+ *   Relations
+ * @property-read \Illuminate\Database\Eloquent\Collection<\Tests\Utils\Models\Post> $post
+ */
+
+final class PostStatus extends Model
+{
+    /** @return \Illuminate\Database\Eloquent\Relations\HasOne<Post> */
```
Comment (spawnia) [inline tests/Utils/Models/PostStatus.php:23]: use fqcn

File: tests/Utils/Models/Task.php [hunk 5]
```diff
@@ -132,4 +133,16 @@ public function userLoaded(): bool
     {
         return $this->relationLoaded('user');
     }
+
+
+    /** @return \Illuminate\Database\Eloquent\Relations\HasOneThrough<PostStatus> */
+
```
Comment (spawnia) [inline tests/Utils/Models/Task.php:139]: redundant newlines

File: tests/database/migrations/2024_07_15_000009_create_testbench_post_statuses_table.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+use Illuminate\Support\Facades\Schema;
+
+final class CreateTestbenchPostStatusesTable extends Migration
+{
+    public function up(): void
+    {
+        Schema::create('post_statuses', function (Blueprint $table): void {
+            $table->increments('id');
+            $table->enum('status' , ["DONE" , "PENDING"]);
+            $table->unsignedInteger("post_id")
```
Comment (spawnia) [inline tests/database/migrations/2024_07_15_000009_create_testbench_post_statuses_table.php:14]: why mix single quotes and double quotes?

File: tests/Integration/Schema/Directives/HasOneThroughDirectiveTest.php
Comment (spawnia) [inline tests/Integration/Schema/Directives/HasOneThroughDirectiveTest.php:1]: Many PHPStan errors still here.

**Review by spawnia — CHANGES_REQUESTED** [review]

**Comment by saeed-rostami** [comment]
> Please try to use the existing test models instead of adding new ones. The addition of 3 new test models, factories and migrations adds quite a bit of maintenance burden.



> Please try to use the existing test models instead of adding new ones. The addition of 3 new test models, factories and migrations adds quite a bit of maintenance burden.



> Please try to use the existing test models instead of adding new ones. The addition of 3 new test models, factories and migrations adds quite a bit of maintenance burden.

ok, I deleted three new models and migrations and used Task and Post models which exist, but I had to add a new one (PostStatus) because there were not three models with one-to-one relation...Task -> Post -> PostStatus

File: tests/database/migrations/2024_07_15_000009_create_testbench_post_statuses_table.php (hunk 6 above)
Comment (saeed-rostami) [reply tests/database/migrations/2024_07_15_000009_create_testbench_post_statuses_table.php]: fixed

File: tests/Utils/Models/Task.php (hunk 5 above)
Comment (saeed-rostami) [reply tests/Utils/Models/Task.php]: removed

File: tests/Utils/Models/PostStatus.php (hunk 4 above)
Comment (saeed-rostami) [reply tests/Utils/Models/PostStatus.php]: added

File: tests/Utils/Models/Post.php (hunk 3 above)
Comment (saeed-rostami) [reply tests/Utils/Models/Post.php]: added

File: tests/Utils/Models/Post.php (hunk 2 above)
Comment (saeed-rostami) [reply tests/Utils/Models/Post.php]: removed

File: tests/Integration/Schema/Directives/HasOneThroughDirectiveTest.php
Comment (saeed-rostami) [reply tests/Integration/Schema/Directives/HasOneThroughDirectiveTest.php:1]: passed

**Comment by spawnia** [comment]
Released with https://github.com/nuwave/lighthouse/releases/tag/v6.42.0.

**Pull request merged**